lib/src/charlatan.dart

```diff
@@ -11,14 +12,37 @@ import 'package:charlatan/src/charlatan_http_response_definition.dart';
 /// ```
 /// {@endtemplate}
 class Charlatan {
-  final Map<String, List<CharlatanHttpResponseDefinition>> _mapping = {};
-
   /// {@nodoc}
   bool shouldLogErrors = true;
 
   /// {@nodoc}
   void silenceErrors() => shouldLogErrors = false;
 
+  final List<CharlatanResponseDefinition> _matchers = [];
+
+  /// Adds a fake response definition for a request that matches the provided
+  /// [requestMatcher]. If the response is not a [CharlatanHttpResponse] then
+  /// the [statusCode] will be used.
+  ///
+  /// [description] is used to describe the response for debugging; it defaults
+  /// to 'Custom Matcher' but should be provided for clarity, e.g. 'GET /users/123?q=foo'
+  void whenMatch(
+    CharlatanRequestMatcher requestMatcher,
+    CharlatanResponseBuilder responseBuilder, {
+    int statusCode = 200,
+    String? description,
+  }) {
+    _matchers.insert(
+      0,
+      CharlatanResponseDefinition(
+        description: description ?? 'Custom Matcher',
+        requestMatcher: requestMatcher,
+        responseBuilder: responseBuilder,
+        defaultStatusCode: statusCode,
+      ),
+    );
+  }
+
```

Comment:
for some more understanding of where this is valuable...

now we can support matching against graphql API calls like this:

```dart
c.whenMatch(
  (r) => r.path == '/api/graphql' && r.body.asJson['queryName'] == 'GetUserQuery',
  (r) => {'name': 'bilbo'}, // but obviously in the shape of GQL responses
);
```

which we can then wrap in some helpers to look more like this:

```dart
c.whenMatch(
  charlatanMatchesGqlQuery('GetUserQuery'),
  charlatanGqlResponse({'name': 'bilbo'}),
);
```